### ISSUE TITLE
Failing Tests with PHP Strict standards messages and Fatal errors

### ISSUE THREAD

**Issue by mhauri** [magento/magento2#2639]
After Installing Magento 2 the Integrator way (works perfect btw.) I tried to run the tests, but it failed with Strict notices and Fatal errors.

``` bash
php bin/magento dev:tests:run unit

---- /dev/magento2/dev/tests/unit> php /dev/magento2/./vendor/phpunit/phpunit/phpunit  

PHP Strict standards:  Declaration of Magento\Newsletter\Test\Unit\Model\Queue\TransportBuilderTest::testGetTransport() should be compatible with Magento\Framework\Mail\Test\Unit\Template\TransportBuilderTest::testGetTransport($templateType, $messageType, $bodyText, $templateNamespace) in /dev/magento2/vendor/magento/module-newsletter/Test/Unit/Model/Queue/TransportBuilderTest.php on line 12
PHP Stack trace:
PHP   1. {main}() /dev/magento2/vendor/phpunit/phpunit/phpunit:0
PHP   2. PHPUnit_TextUI_Command::main() /dev/magento2/vendor/phpunit/phpunit/phpunit:55
PHP   3. PHPUnit_TextUI_Command->run() /dev/magento2/vendor/phpunit/phpunit/src/TextUI/Command.php:132
PHP   4. PHPUnit_TextUI_Command->handleArguments() /dev/magento2/vendor/phpunit/phpunit/src/TextUI/Command.php:141
PHP   5. PHPUnit_Util_Configuration->getTestSuiteConfiguration() /dev/magento2/vendor/phpunit/phpunit/src/TextUI/Command.php:681
PHP   6. PHPUnit_Util_Configuration->getTestSuite() /dev/magento2/vendor/phpunit/phpunit/src/Util/Configuration.php:825
PHP   7. PHPUnit_Framework_TestSuite->addTestFiles() /dev/magento2/vendor/phpunit/phpunit/src/Util/Configuration.php:911
PHP   8. PHPUnit_Framework_TestSuite->addTestFile() /dev/magento2/vendor/phpunit/phpunit/src/Framework/TestSuite.php:411
PHP   9. PHPUnit_Util_Fileloader::checkAndLoad() /dev/magento2/vendor/phpunit/phpunit/src/Framework/TestSuite.php:358
PHP  10. PHPUnit_Util_Fileloader::load() /dev/magento2/vendor/phpunit/phpunit/src/Util/Fileloader.php:77
PHP  11. include_once() /dev/magento2/vendor/phpunit/phpunit/src/Util/Fileloader.php:93

Strict standards: Declaration of Magento\Newsletter\Test\Unit\Model\Queue\TransportBuilderTest::testGetTransport() should be compatible with Magento\Framework\Mail\Test\Unit\Template\TransportBuilderTest::testGetTransport($templateType, $messageType, $bodyText, $templateNamespace) in /dev/magento2/vendor/magento/module-newsletter/Test/Unit/Model/Queue/TransportBuilderTest.php on line 12

Call Stack:
    0.0005     228448   1. {main}() /dev/magento2/vendor/phpunit/phpunit/phpunit:0
    0.0121     941888   2. PHPUnit_TextUI_Command::main() /dev/magento2/vendor/phpunit/phpunit/phpunit:55
    0.0121     942504   3. PHPUnit_TextUI_Command->run() /dev/magento2/vendor/phpunit/phpunit/src/TextUI/Command.php:132
    0.0121     945040   4. PHPUnit_TextUI_Command->handleArguments() /dev/magento2/vendor/phpunit/phpunit/src/TextUI/Command.php:141
    0.0224    1802760   5. PHPUnit_Util_Configuration->getTestSuiteConfiguration() /dev/magento2/vendor/phpunit/phpunit/src/TextUI/Command.php:681
    0.0225    1804024   6. PHPUnit_Util_Configuration->getTestSuite() /dev/magento2/vendor/phpunit/phpunit/src/Util/Configuration.php:825
    1.2586   32599200   7. PHPUnit_Framework_TestSuite->addTestFiles() /dev/magento2/vendor/phpunit/phpunit/src/Util/Configuration.php:911
    9.6873  263754088   8. PHPUnit_Framework_TestSuite->addTestFile() /dev/magento2/vendor/phpunit/phpunit/src/Framework/TestSuite.php:411
    9.6880  264244680   9. PHPUnit_Util_Fileloader::checkAndLoad() /dev/magento2/vendor/phpunit/phpunit/src/Framework/TestSuite.php:358
    9.6880  264244864  10. PHPUnit_Util_Fileloader::load() /dev/magento2/vendor/phpunit/phpunit/src/Util/Fileloader.php:77
    9.6887  264289584  11. include_once('/dev/magento2/vendor/magento/module-newsletter/Test/Unit/Model/Queue/TransportBuilderTest.php') /dev/magento2/vendor/phpunit/phpunit/src/Util/Fileloader.php:93

PHPUnit 4.1.0 by Sebastian Bergmann.

Configuration read from /dev/magento2/dev/tests/unit/phpunit.xml.dist

...........................................................    59 / 16279 (  0%)
...........................................................   118 / 16279 (  0%)
...........................................................   177 / 16279 (  1%)
...........................................................   236 / 16279 (  1%)
...........................................................   295 / 16279 (  1%)
...........................................................   354 / 16279 (  2%)
...........................................................   413 / 16279 (  2%)
...........................................................   472 / 16279 (  2%)
...........................................................   531 / 16279 (  3%)
...........................................................   590 / 16279 (  3%)
...........................................................   649 / 16279 (  3%)
...........................................................   708 / 16279 (  4%)
...........................................................   767 / 16279 (  4%)
...........................................................   826 / 16279 (  5%)
...........................................................   885 / 16279 (  5%)
...........................................................   944 / 16279 (  5%)
...........................................................  1003 / 16279 (  6%)
...........................................................  1062 / 16279 (  6%)
...........................................................  1121 / 16279 (  6%)
...........................................................  1180 / 16279 (  7%)
...........................................................  1239 / 16279 (  7%)
...........................................................  1298 / 16279 (  7%)
...........................................................  1357 / 16279 (  8%)
...........................................................  1416 / 16279 (  8%)
...........................................................  1475 / 16279 (  9%)
...........................................................  1534 / 16279 (  9%)
...........................................................  1593 / 16279 (  9%)
...........................................................  1652 / 16279 ( 10%)
...........................................................  1711 / 16279 ( 10%)
...........................................................  1770 / 16279 ( 10%)
...........................................................  1829 / 16279 ( 11%)
...........................................................  1888 / 16279 ( 11%)
...........................................................  1947 / 16279 ( 11%)
...........................................................  2006 / 16279 ( 12%)
...........................................................  2065 / 16279 ( 12%)
...................................................SSSS....  2124 / 16279 ( 13%)
...........................................................  2183 / 16279 ( 13%)
...........................................................  2242 / 16279 ( 13%)
...........................................................  2301 / 16279 ( 14%)
...........................................................  2360 / 16279 ( 14%)
...........................................................  2419 / 16279 ( 14%)
...........................................................  2478 / 16279 ( 15%)
...........................................................  2537 / 16279 ( 15%)
...........................................................  2596 / 16279 ( 15%)
...........................................................  2655 / 16279 ( 16%)
...........................................................  2714 / 16279 ( 16%)
...........................................................  2773 / 16279 ( 17%)
...........................................................  2832 / 16279 ( 17%)
...........................................................  2891 / 16279 ( 17%)
...........................................................  2950 / 16279 ( 18%)
...........................................................  3009 / 16279 ( 18%)
.........................................I.I...............  3068 / 16279 ( 18%)
...........................................................  3127 / 16279 ( 19%)
...........................................................  3186 / 16279 ( 19%)
...........................................................  3245 / 16279 ( 19%)
...........................................................  3304 / 16279 ( 20%)
...........................................................  3363 / 16279 ( 20%)
...........................................................  3422 / 16279 ( 21%)
...........................................................  3481 / 16279 ( 21%)
...........................................................  3540 / 16279 ( 21%)
...........................................................  3599 / 16279 ( 22%)
...........................................................  3658 / 16279 ( 22%)
...........................................................  3717 / 16279 ( 22%)
...........................................................  3776 / 16279 ( 23%)
...........................................................  3835 / 16279 ( 23%)
...........................................................  3894 / 16279 ( 23%)
...........................................................  3953 / 16279 ( 24%)
...........................................................  4012 / 16279 ( 24%)
...........................................................  4071 / 16279 ( 25%)
...........................................................  4130 / 16279 ( 25%)
...........................................................  4189 / 16279 ( 25%)
...........................................................  4248 / 16279 ( 26%)
...........................................................  4307 / 16279 ( 26%)
...........................................................  4366 / 16279 ( 26%)
...........................................................  4425 / 16279 ( 27%)
...........................................................  4484 / 16279 ( 27%)
...........................................................  4543 / 16279 ( 27%)
...........................................................  4602 / 16279 ( 28%)
...........................................................  4661 / 16279 ( 28%)
...........................................................  4720 / 16279 ( 28%)
...........................................................  4779 / 16279 ( 29%)
...........................................................  4838 / 16279 ( 29%)
...........................................................  4897 / 16279 ( 30%)
..................................................SSSSSSSS.  4956 / 16279 ( 30%)
...........................................................  5015 / 16279 ( 30%)
...........................................................  5074 / 16279 ( 31%)
......................................................S....  5133 / 16279 ( 31%)
...........................................................  5192 / 16279 ( 31%)
...........................................................  5251 / 16279 ( 32%)
...........................................................  5310 / 16279 ( 32%)
...........................................................  5369 / 16279 ( 32%)
...........................................................  5428 / 16279 ( 33%)
...........................................................  5487 / 16279 ( 33%)
...........................................................  5546 / 16279 ( 34%)
...........................................................  5605 / 16279 ( 34%)
...........................................................  5664 / 16279 ( 34%)
...........................................................  5723 / 16279 ( 35%)
...........................................................  5782 / 16279 ( 35%)
...........................................................  5841 / 16279 ( 35%)
...........................................................  5900 / 16279 ( 36%)
...........................................................  5959 / 16279 ( 36%)
..................SSSSSSSSSSSSSSSSSSSSS....................  6018 / 16279 ( 36%)
...........................................................  6077 / 16279 ( 37%)
...........................................................  6136 / 16279 ( 37%)
...........................................................  6195 / 16279 ( 38%)
...........................PHP Fatal error:  Call to a member function dispatch() on null in /dev/magento2/vendor/magento/framework/Model/AbstractModel.php on line 626
PHP Stack trace:
PHP   1. {main}() /dev/magento2/vendor/phpunit/phpunit/phpunit:0
PHP   2. PHPUnit_TextUI_Command::main() /dev/magento2/vendor/phpunit/phpunit/phpunit:55
PHP   3. PHPUnit_TextUI_Command->run() /dev/magento2/vendor/phpunit/phpunit/src/TextUI/Command.php:132
PHP   4. PHPUnit_TextUI_TestRunner->doRun() /dev/magento2/vendor/phpunit/phpunit/src/TextUI/Command.php:179
PHP   5. PHPUnit_Framework_TestSuite->run() /dev/magento2/vendor/phpunit/phpunit/src/TextUI/TestRunner.php:426
PHP   6. PHPUnit_Framework_TestSuite->run() /dev/magento2/vendor/phpunit/phpunit/src/Framework/TestSuite.php:675
PHP   7. PHPUnit_Framework_TestCase->run() /dev/magento2/vendor/phpunit/phpunit/src/Framework/TestSuite.php:675
PHP   8. PHPUnit_Framework_TestResult->run() /dev/magento2/vendor/phpunit/phpunit/src/Framework/TestCase.php:753
PHP   9. PHPUnit_Framework_TestCase->runBare() /dev/magento2/vendor/phpunit/phpunit/src/Framework/TestResult.php:686
PHP  10. PHPUnit_Framework_TestCase->runTest() /dev/magento2/vendor/phpunit/phpunit/src/Framework/TestCase.php:817
PHP  11. ReflectionMethod->invokeArgs() /dev/magento2/vendor/phpunit/phpunit/src/Framework/TestCase.php:951
PHP  12. Magento\Eav\Test\Unit\Model\ResourceModel\Entity\AttributeTest->testSaveOptionNewUserDefinedAttribute() /dev/magento2/vendor/phpunit/phpunit/src/Framework/TestCase.php:951
PHP  13. Magento\Framework\Model\ResourceModel\Db\AbstractDb->save() /dev/magento2/vendor/magento/module-eav/Test/Unit/Model/ResourceModel/Entity/AttributeTest.php:195
PHP  14. Magento\Framework\Model\ResourceModel\AbstractResource->commit() /dev/magento2/vendor/magento/framework/Model/ResourceModel/Db/AbstractDb.php:410
PHP  15. call_user_func:{/dev/magento2/vendor/magento/framework/Model/ResourceModel/AbstractResource.php:94}() /dev/magento2/vendor/magento/framework/Model/ResourceModel/AbstractResource.php:94
PHP  16. Magento\Framework\Model\AbstractModel->afterCommitCallback() /dev/magento2/vendor/magento/framework/Model/ResourceModel/AbstractResource.php:94

Fatal error: Call to a member function dispatch() on null in /dev/magento2/vendor/magento/framework/Model/AbstractModel.php on line 626

Call Stack:
    0.0005     228448   1. {main}() /dev/magento2/vendor/phpunit/phpunit/phpunit:0
    0.0121     941888   2. PHPUnit_TextUI_Command::main() /dev/magento2/vendor/phpunit/phpunit/phpunit:55
    0.0121     942504   3. PHPUnit_TextUI_Command->run() /dev/magento2/vendor/phpunit/phpunit/src/TextUI/Command.php:132
   30.0017  534351704   4. PHPUnit_TextUI_TestRunner->doRun() /dev/magento2/vendor/phpunit/phpunit/src/TextUI/Command.php:179
   30.0046  534743880   5. PHPUnit_Framework_TestSuite->run() /dev/magento2/vendor/phpunit/phpunit/src/TextUI/TestRunner.php:426
  131.1296 1066943152   6. PHPUnit_Framework_TestSuite->run() /dev/magento2/vendor/phpunit/phpunit/src/Framework/TestSuite.php:675
  131.2897 1066577216   7. PHPUnit_Framework_TestCase->run() /dev/magento2/vendor/phpunit/phpunit/src/Framework/TestSuite.php:675
  131.2898 1066577216   8. PHPUnit_Framework_TestResult->run() /dev/magento2/vendor/phpunit/phpunit/src/Framework/TestCase.php:753
  131.2898 1066578032   9. PHPUnit_Framework_TestCase->runBare() /dev/magento2/vendor/phpunit/phpunit/src/Framework/TestResult.php:686
  131.2908 1066750792  10. PHPUnit_Framework_TestCase->runTest() /dev/magento2/vendor/phpunit/phpunit/src/Framework/TestCase.php:817
  131.2908 1066751688  11. ReflectionMethod->invokeArgs() /dev/magento2/vendor/phpunit/phpunit/src/Framework/TestCase.php:951
  131.2908 1066752272  12. Magento\Eav\Test\Unit\Model\ResourceModel\Entity\AttributeTest->testSaveOptionNewUserDefinedAttribute() /dev/magento2/vendor/phpunit/phpunit/src/Framework/TestCase.php:951
  131.2966 1066897432  13. Magento\Framework\Model\ResourceModel\Db\AbstractDb->save() /dev/magento2/vendor/magento/module-eav/Test/Unit/Model/ResourceModel/Entity/AttributeTest.php:195
  131.3446 1067675080  14. Magento\Framework\Model\ResourceModel\AbstractResource->commit() /dev/magento2/vendor/magento/framework/Model/ResourceModel/Db/AbstractDb.php:410
  131.3482 1067732232  15. call_user_func:{/dev/magento2/vendor/magento/framework/Model/ResourceModel/AbstractResource.php:94}() /dev/magento2/vendor/magento/framework/Model/ResourceModel/AbstractResource.php:94
  131.3482 1067732344  16. Magento\Framework\Model\AbstractModel->afterCommitCallback() /dev/magento2/vendor/magento/framework/Model/ResourceModel/AbstractResource.php:94


---- /dev/magento2/dev/tests/static/framework/tests/unit> php /dev/magento2/./vendor/phpunit/phpunit/phpunit  

PHPUnit 4.1.0 by Sebastian Bergmann.

Configuration read from /dev/magento2/dev/tests/static/framework/tests/unit/phpunit.xml.dist

...............................F...F......FFF.................... 65 / 66 ( 98%)
.

Time: 444 ms, Memory: 11.25Mb

There were 5 failures:

1) Magento\TestFramework\Dependency\PhpRuleTest::testGetDependencyInfo with data set "Extend class in different module" ('Magento\\AnotherModule\\SomeClass', 'something extends \\Magento\\SomeModule\\Any\\ClassName {', array(array('Magento\\SomeModule', 'hard', 'Magento\\SomeModule\\Any\\ClassName')))
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
-    0 => Array (...)
 )

/dev/magento2/dev/tests/static/framework/tests/unit/testsuite/Magento/TestFramework/Dependency/PhpRuleTest.php:42

2) Magento\TestFramework\Dependency\PhpRuleTest::testGetDependencyInfo with data set "Helper class from another module" ('Magento\\AnotherModule\\SomeClass', '$this->helper("Magento\\SomeModule\\Any\\ClassName")', array(array('Magento\\SomeModule', 'hard', 'Magento\\SomeModule\\Any\\ClassName')))
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
-    0 => Array (...)
 )

/dev/magento2/dev/tests/static/framework/tests/unit/testsuite/Magento/TestFramework/Dependency/PhpRuleTest.php:42

3) Magento\TestFramework\Dependency\PhpRuleTest::testGetDependencyInfo with data set "Plugin on class in different module" ('Magento\\Module1\\Plugin2', 'Magento\\Module2\\Subject', array(array('Magento\\Module2', 'soft', 'Magento\\Module2\\Subject')))
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
-    0 => Array (...)
 )

/dev/magento2/dev/tests/static/framework/tests/unit/testsuite/Magento/TestFramework/Dependency/PhpRuleTest.php:42

4) Magento\TestFramework\Dependency\PhpRuleTest::testGetDependencyInfo with data set "Plugin depends on arbitrary class in same module as subject" ('Magento\\Module1\\Plugin2', 'Magento\\Module2\\NotSubject', array(array('Magento\\Module2', 'soft', 'Magento\\Module2\\NotSubject')))
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
-    0 => Array (...)
 )

/dev/magento2/dev/tests/static/framework/tests/unit/testsuite/Magento/TestFramework/Dependency/PhpRuleTest.php:42

5) Magento\TestFramework\Dependency\PhpRuleTest::testGetDependencyInfo with data set "Plugin depends on arbitrary class in arbitrary module" ('Magento\\Module1\\Plugin2', 'Magento\\OtherModule\\NotSubject', array(array('Magento\\OtherModule', 'hard', 'Magento\\OtherModule\\NotSubject')))
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
-    0 => Array (...)
 )

/dev/magento2/dev/tests/static/framework/tests/unit/testsuite/Magento/TestFramework/Dependency/PhpRuleTest.php:42

FAILURES!                              
Tests: 66, Assertions: 89, Failures: 5.

---- /dev/magento2/dev/tests/integration/framework/tests/unit> php /dev/magento2/./vendor/phpunit/phpunit/phpunit  

PHPUnit 4.1.0 by Sebastian Bergmann.

Configuration read from /dev/magento2/dev/tests/integration/framework/tests/unit/phpunit.xml.dist

...............................................................  63 / 164 ( 38%)
............................................................... 126 / 164 ( 76%)
.................S....................

Time: 1.01 seconds, Memory: 21.50Mb

OK, but incomplete, skipped, or risky tests!
Tests: 164, Assertions: 318, Skipped: 1.    
----------------------------------------------------------------------
FAILED - 2 of 3:
 - /dev/magento2/dev/tests/unit> php /dev/magento2/./vendor/phpunit/phpunit/phpunit 
 - /dev/magento2/dev/tests/static/framework/tests/unit> php /dev/magento2/./vendor/phpunit/phpunit/phpunit 
```

**Comment by mazhalai** [comment]
@mhauri Thank you for reporting, we have created MAGETWO-46658 to fix it.

**Comment by daim2k5** [comment]
@mhauri maybe you can test if phpunit 5.1 runs better?

**Comment by Vinai** [comment]
With PHPUnit 5 things have become more strict. Also, 5.1 has some [breaking changes](https://twitter.com/s_bergmann/status/673881844614582272), and phpunit mock object < 4.0.3 has some problems with stubbing `__toString()`, so all in all I think the best way forward would be to actually fix the tests on a one by one basis.
Then, once all are green, an upgrade might be considered.

**Comment by vkholoshenko** [comment]
Hi, @mhauri! The fix has been merged with 33b13b65bfbd6d32bc4a70d15e9f1bf4aab1f695. Thank you for helping us make Magento better!
